pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2019. PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// +k8s:openapi-gen=true
+// TidbMonitor encode the spec and status of the monitoring component of a TiDB cluster
+type TidbMonitor struct {
+	metav1.TypeMeta `json:",inline"`
+	// +k8s:openapi-gen=false
+	metav1.ObjectMeta `json:"metadata"`
+
+	// Spec defines the desired state of TidbMonitor
+	Spec TidbMonitorSpec `json:"spec"`
+
+	// +k8s:openapi-gen=false
+	// Most recently observed status of the TidbMonitor
+	Status TidbMonitorStatus `json:"status"`
+}
+
+// +k8s:openapi-gen=true
+// TidbMonitor spec encode the desired state of tidb monitoring component
+type TidbMonitorSpec struct {
+	Clusters []TidbClusterRef `json:"clusters"`
+
+	Prometheus PrometheusSpec `json:"prometheus"`
+	// +optional
+	Grafana     *GrafanaSpec    `json:"grafana"`
+	Reloader    ReloaderSpec    `json:"reloader"`
+	Initializer InitializerSpec `json:"initializer"`
+
+	ImagePullPolicy corev1.PullPolicy `json:"imagePullPolicy,omitempty"`
+	// +optional
+	Persistent bool `json:"persistent,omitempty"`
+	// +optional
+	StorageClassName string `json:"storageClassName,omitempty"`
+	// +optional
+	Storage string `json:"storage,omitempty"`
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+	// +optional
+	Annotations map[string]string `json:"annotations,omitempty"`
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+
+	// kubePrometheusURL is where tidb-monitoring get the  common metrics of kube-prometheus.
+	// Ref: https://github.com/coreos/kube-prometheus
+	// +optional
+	KubePrometheusURL string `json:"kubePrometheusURL,omitempty"`
```

Comment:
Should we define these optional fields as pointers?

Reply:
It depends on whether empty string (`""`) should be considered as "unset" for this field (typically, field cannot be unset when we merge changes to existing object, e.g. server-side apply), THIS one should is better to be a pointer I guess, I will check whether other `string` fields should be changed to `*string`